### PR TITLE
manually implement traits for Span

### PR DESCRIPTION
**Pull request by lcnr**
`PartialEq::eq` is used a few billion times, so it might be worth it to go the extra mile here.

This also changes `base_or_index` to a `newtype_index`, meaning that the 256 highest values are used as a niche, thereby allowing `Option<Span>` to have the same size as `Span`.

r? @nnethercote

### CODE REVIEW

File: src/librustc_span/span_encoding.rs [hunk 1]
```diff
@@ -65,13 +65,24 @@ use std::marker::{StructuralEq, StructuralPartialEq};
 #[derive(Clone, Copy)]
 #[repr(C)] // So we can transmute to a `u64`.
 pub struct Span {
-    base_or_index: u32,
+    // We can transmute `SpanBaseOrIndex` to an `u32` as it is internally
```
Comment (nnethercote) [inline src/librustc_span/span_encoding.rs]: Nit: "a `u32`"

File: src/librustc_span/span_encoding.rs [hunk 2]
```diff
@@ -65,13 +65,24 @@ use std::marker::{StructuralEq, StructuralPartialEq};
...
     ctxt_or_zero: u16,
 }
 
+rustc_index::newtype_index! {
+    pub struct SpanBaseOrIndex {
+        const DUMMY_SPAN_INDEX = 0,
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
```
Comment (nnethercote) [inline src/librustc_span/span_encoding.rs:82]: This is worth a comment. E.g. "`SpanBaseOrIndex` is implemented via `newtype_index`, which means that `Option<Span>` takes the same space as `Span`.

File: src/librustc_span/span_encoding.rs [hunk 3]
```diff
@@ -73,6 +73,7 @@ pub struct Span {
 }
 
 rustc_index::newtype_index! {
+    #[repr(transparent)]
```
Comment (nnethercote) [inline src/librustc_span/span_encoding.rs:76]: Please add a comment explaining why this is present.

File: src/librustc_span/span_encoding.rs [hunk 4]
```diff
@@ -84,8 +85,12 @@ impl PartialEq for Span {
     }
 }
 
+impl StructuralPartialEq for Span {}
```
Comment (nnethercote) [inline src/librustc_span/span_encoding.rs:100]: Please add a comment explaining why this is present.

**Comment by nnethercote** [comment]
There are some nice ideas here! I like the optimized `PartialEq` and the space optimization for `Option<Span>`.

The `StructuralEq` and `repr(transparent)` changes are something of a mystery to me. I looked up what they mean, but I still don't understand their benefit for `Span`.

I have a preference for commit messages that have a full sentence on their first line, and I don't like commits that fix up things that could easily have been fixed in earlier commits (e.g. the "words" commit) so I'd be happy to see some clean-ups there. 

Finally, whether all this is worth it will depend on the perf results, so I will wait for them before giving r+. Do you have any local measurements?

**Comment by Mark-Simulacrum** [comment]
bors died so the build didn't get queued -- @lcnr, do you want to re-run try or we can invoke build on 9c509a323b82b57aae17bc2a2060471600a0307f?

**Comment by lcnr** [comment]
Doesn't really look like an improvement :/

Not sure if we want to go ahead here or just drop this

**Comment by nnethercote** [comment]
Probably better to drop it, because it introduces additional complexity.

**Pull request closed without merging**